components/encoder/as5048a.go

```diff
@@ -0,0 +1,296 @@
+	// by (360 / 2^14) to get degrees
+	byteData := (int(msB) << 6) | int(lsB)
+	return (float64(byteData) * scalingFactor), nil
+}
+
+func (enc *AS5048A) updatePosition(ctx context.Context) error {
+	enc.mu.Lock()
+	defer enc.mu.Unlock()
+	angleDeg, err := enc.readPosition(ctx)
+	if err != nil {
+		return err
+	}
+	angleDeg += enc.positionOffset
+	// in order to keep track of multiple rotations, we increment / decrement
+	// a rotations counter whenever two subsequent positions are on either side
+	// of 0 (or 360) within a window of 2 * transitionEpsilon
+	forwardsTransition := (angleDeg <= transitionEpsilon) && ((360.0 - enc.position) <= transitionEpsilon)
+	backwardsTransition := (enc.position <= transitionEpsilon) && ((360.0 - angleDeg) <= transitionEpsilon)
+	if forwardsTransition {
+		enc.rotations++
+	} else if backwardsTransition {
+		enc.rotations--
+	}
+	enc.position = angleDeg
+	return nil
+}
+
+// TicksCount returns the total number of rotations detected
+// by the encoder (rather than a number of pulse state transitions)
+// because this encoder is absolute and not incremental. As a result
+// a user MUST set ticks_per_rotation on the config of the corresponding
+// motor to 1. Any other value will result in completely incorrect
+// position measurements by the motor.
+func (enc *AS5048A) TicksCount(
+	ctx context.Context, extra map[string]interface{},
+) (float64, error) {
+	enc.mu.RLock()
+	defer enc.mu.RUnlock()
+	ticks := float64(enc.rotations) + enc.position/360.0
+	return ticks, nil
+}
+
+// Reset sets the current position measured by the encoder to be considered
+// its new zero position. If the offset provided is not 0.0, it also
+// sets the positionOffset attribute and adjusts all future recorded
+// positions by that offset (until the function is called again).
+func (enc *AS5048A) Reset(
+	ctx context.Context, offset float64, extra map[string]interface{},
+) error {
+	enc.mu.Lock()
+	defer enc.mu.Unlock()
+	enc.positionOffset = offset
+	enc.position = 0.0 + offset
+	currentMSB, err := enc.i2cHandle.ReadByteData(ctx, byte(0xFE))
+	if err != nil {
+		return err
+	}
+	currentLSB, err := enc.i2cHandle.ReadByteData(ctx, byte(0xFF))
+	if err != nil {
+		return err
+	}
+	// clear current zero position
+	err = enc.i2cHandle.WriteByteData(ctx, byte(0x16), byte(0))
+	if err != nil {
+		return err
+	}
+	err = enc.i2cHandle.WriteByteData(ctx, byte(0x17), byte(0))
+	if err != nil {
+		return err
+	}
+	// write current position to zero register
+	err = enc.i2cHandle.WriteByteData(ctx, byte(0x16), currentMSB)
+	if err != nil {
+		return err
+	}
+	err = enc.i2cHandle.WriteByteData(ctx, byte(0x17), currentLSB)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Close stops the position loop of the encoder when the component
+// is closed.
+func (enc *AS5048A) Close() {
+	enc.cancel()
+	enc.activeBackgroundWorkers.Wait()
+}
+
+// NewAS5048AEncoder returns a new instance of an AS5048A encoder.
+func NewAS5048AEncoder(
```

Comment:
Can you make this private. I prefere having constructor near the top of the file or the objects they're creating, but no biggie if you prefer it here.